frontends/parsers/p4/p4parser.ypp

```diff
@@ -0,0 +1,1012 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Use Bison's C++ mode.
+%skeleton "lalr1.cc" /* -*-C++-*- */
+%require "3.0.4"
+
+// Set up names.
+%defines
+%define api.namespace {P4}
+%define parser_class_name {P4Parser}
+
+// Use the C++-native variant representation of tokens.
+%define api.token.constructor
+%define api.value.type variant
+
+// Add asserts to ensure that we use the API correctly.
+%define parse.assert true
+
+// Enable verbose error reporting.
+%error-verbose
+
+// Declare dependencies.
+%code requires {
+namespace P4 {
+class P4Lexer;
+class P4ParserDriver;
+}  // namespace P4
+
+// Bison uses the types you provide to %type to make constructors for the
+// variant type it uses under the hood, but its code generation is a little
+// naive and it always prepends 'const' to the type. This is problematic when
+// the symbol type we want is itself const, since duplicate const qualifiers are
+// forbidden in C++. We avoid the problem using a typedef.
+typedef const IR::Type ConstType;
+
+#ifndef YYDEBUG
+#define YYDEBUG 1
+#endif
+
+#define YY_NULLPTR nullptr
+
+#include "frontends/common/constantParsing.h"
+#include "lib/cstring.h"
+#include "lib/source_file.h"
+}
+
+%param { P4::P4ParserDriver& driver }
+%parse-param { P4::P4Lexer& lexer }
+
+// Use location tracking with our custom location type.
+%locations
+%define api.location.type {Util::SourceInfo}
+
+%{ /* -*-C++-*- */
+#include <cassert>   // NOLINT(build/include_order)
+#include <iostream>  // NOLINT(build/include_order)
+
+#include "frontends/parsers/parserDriver.h"
+#include "frontends/parsers/p4/p4lexer.hpp"
+#include "frontends/parsers/p4/p4parser.hpp"
+#include "ir/ir.h"
+
+#define YYLLOC_DEFAULT(Cur, Rhs, N)                                     \
+    ((Cur) = (N) ? YYRHSLOC(Rhs, 1) + YYRHSLOC(Rhs, N)                  \
+                 : Util::SourceInfo(YYRHSLOC(Rhs, 0).getEnd()))
+
+#undef yylex
+#define yylex lexer.yylex
+
+%}
+
+// Use iostreams to print all values.
+%printer { yyoutput << $$; } <*>;
+
+%define api.token.prefix {TOK_}
+%token      END
+%token      UNEXPECTED_TOKEN
+%token      LE "<="
+%token      GE ">="
+%token      SHL "<<"
+%token      AND "&&"
+%token      OR "||"
+%token      NE "!="
+%token      EQ "=="
+%token      PLUS "+"
+%token      MINUS "-"
+%token      MUL "*"
+%token      DIV "/"
+%token      MOD "%"
+%token      BIT_OR "|"
+%token      BIT_AND "&"
+%token      BIT_XOR "^"
+%token      COMPLEMENT "~"
+%token      L_BRACKET "["
+%token      R_BRACKET "]"
+%token      L_BRACE "{"
```

Comment:
It's actually possible to use the aliases that are defined here in the grammar; for example, we could refer to L_BRACE as "{". I waffled on whether to do that, and I ultimately went this direction because it keeps things consistent between the lexer and the parser. I'm happy to change it, though, since it sounds like you'd prefer to do it the other way.